### PR TITLE
Add ClassDecl/Type import functionality.

### PR DESCRIPTION
**Pull request by jonmeow**
I believe this PR is sufficient to pull in all current class features, including the current bits of inheritance which have been implemented. Because a class declaration can reference its own type, this creates an incomplete type prior to constant loading.

Right now, the object representation is imported proactively, but individual fields are left as ImportRefUnused. This means that member functions and similar will only be imported if called.

This also adjusts how function parameters are being handled, to match the expectations of Self param structure.

When formatting, I'm starting to look into constants. Otherwise we get "unexpected instref".

Overall, there are a few things that may be worth further discussion:

- The lack of a constant corresponding to the ClassType on ClassDecl is inconvenient -- I'd like to see how zygoloid feels about trying to restructure this. i.e., I'm setting a constant in order to be able to track things down later, it'd be nice if the normal IR did this simply for consistency, or if we were able to combine these rather than having separate instructions.
- Should we shift the parse node tracking further, and go with a setup wherein imports can embed import references into that? e.g., negative values go to another array which includes a ImportIRId for printing diagnostics, replacing the invalid NodeId.
- Can the formatter switch to a more general scan of instructions for naming, to eliminate the ImportRef constant approach added here?
- GetExprValueForLookupResult special-casing instructions felt surprising, I might see if there's a way to restructure to avoid that.

But I think these issues are things that can be separated out.

### CODE REVIEW

File: toolchain/check/handle_name.cpp
Comment (chandlerc) [inline toolchain/check/handle_name.cpp:77]: Any reason not to recurse immediately and regardless of `import_kind`, and rely on the recursing `GetExprValueForLookupResult` to diagnose any unsupported kind? For me, that would make the logic simpler of first recurse to the leaf IR, and then have a single place where we handle the supported leaves or diagnose others.

Comment (chandlerc) [reply toolchain/check/handle_name.cpp:77]: But maybe I've misunderstood how this is working? Let me know if so.

File: toolchain/check/import_ref.cpp
Comment (chandlerc) [inline toolchain/check/import_ref.cpp:182]: Not blocking at all, but curious if you have anything in mind to avoid the parallel code structure?

File: toolchain/check/import_ref.cpp
Comment (chandlerc) [inline toolchain/check/import_ref.cpp]: From reading the rest of the PR, I think I have some idea of how all this is used, but I think it's a somewhat hard to piece together. Could you add some comments, maybe on the `Work` type rather than just this field that try to pull together how this is used throughout the resolution?

File: toolchain/check/import_ref.cpp
Comment (chandlerc) [inline toolchain/check/import_ref.cpp]: A bit more comments to help explain the initial work logic here would help me understand the flow of the function I think.

File: toolchain/check/import_ref.cpp
Comment (chandlerc) [inline toolchain/check/import_ref.cpp:384]: Would it make sense to extract some of this logic into a helper function similar to `CreateIncompleteClass` and leave this to just wire things together?

File: toolchain/check/testdata/class/fail_todo_import_forward_decl.carbon
Comment (chandlerc) [inline toolchain/check/testdata/class/fail_todo_import_forward_decl.carbon]: Any particular reason to switch to the raw dump here?

File: toolchain/sem_ir/formatter.cpp
Comment (chandlerc) [inline toolchain/sem_ir/formatter.cpp:35]: I'm not seeing this used in the tests, am I misunderstanding how this works and it doesn't show up like `file` and such?

**Review by chandlerc — COMMENTED** [review]
Some comments inline!

File: toolchain/check/import_ref.cpp
Comment (jonmeow) [reply toolchain/check/import_ref.cpp:182]: Most of my thoughts are along the lines of either flattening out `addr` (it's a bit of an odd approach right now, maybe bind names need to have more of a modifier concept) + making the presence of a bind name reliable (removing a conditional), or delving into the _other_ end and rewriting param handling to be more agnostic (but that's pretty substantial). It's also possible some combination of both would work best.

[added a TODO; note, I'm thinking about this, but "Figure out" is an implementation comment, not a TODO]

File: toolchain/check/testdata/class/fail_todo_import_forward_decl.carbon
Comment (jonmeow) [reply toolchain/check/testdata/class/fail_todo_import_forward_decl.carbon]: Oops, this was debug. I'd just missed cleaning this up.

File: toolchain/sem_ir/formatter.cpp
Comment (jonmeow) [reply toolchain/sem_ir/formatter.cpp:35]: It's visible in template constants. In something such as:

```
// CHECK:STDOUT: %B.ref.loc6: <function> = name_ref B, file.%import_ref [template = imports.%B]
```

The reason `imports.` is present there is due to this.

Note, there may be some better approach to this, as noted in the PR description ("Can the formatter switch to a more general scan of instructions for naming, to eliminate the ImportRef constant approach added here?").

Added an explicit TODO.

File: toolchain/check/handle_name.cpp
Comment (jonmeow) [reply toolchain/check/handle_name.cpp:77]: Added a comment. Note that in the "else" case it's returning the ImportRefUsed as the value, so a recursion is not equivalent logic.

File: toolchain/check/import_ref.cpp
Comment (jonmeow) [reply toolchain/check/import_ref.cpp]: Added some comments here, moved the comments on TryResolveInst up to the class level and elaborated on them.

File: toolchain/check/import_ref.cpp
Comment (jonmeow) [reply toolchain/check/import_ref.cpp]: I'm not adding anything specific here, but do the other comment changes address this?

File: toolchain/check/import_ref.cpp
Comment (jonmeow) [reply toolchain/check/import_ref.cpp:384]: Done

File: toolchain/check/import_ref.cpp
Comment (chandlerc) [inline toolchain/check/import_ref.cpp]: I'm not sure I'm understanding this sentence...

Do you mean `Import the type with its associated constant`?

Something else?

File: toolchain/check/import_ref.cpp
Comment (chandlerc) [inline toolchain/check/import_ref.cpp:44]: I think this is missing an indent?
```suggestion
//      - For trivial cases with zero or one input constants, this may return
//        a constant (if one, potentially Invalid) directly.
```

File: toolchain/check/import_ref.cpp
Comment (chandlerc) [inline toolchain/check/import_ref.cpp]: Maybe a bit more reminder of the connection between the specific bool variable and the passes?

```suggestion
    // On the first pass, we build the incomplete type's constant above. If we
    // get here on a subsequent pass we need to fetch the one we built in the
    // first pass.
```

File: toolchain/check/testdata/class/import_member_cycle.carbon
Comment (chandlerc) [inline toolchain/check/testdata/class/import_member_cycle.carbon]: Another raw dump maybe for debugging?

File: toolchain/check/import_ref.cpp
Comment (chandlerc) [inline toolchain/check/import_ref.cpp]: What do you think about calling this `resolved_incomplete_type` to make it a bit more clear how it is used? See also the naming suggestion below.

Also happy with other names if a different approach would work.

File: toolchain/check/import_ref.cpp
Comment (chandlerc) [inline toolchain/check/import_ref.cpp]: If the `Work` member is renamed, I'd rename this to match (`resolved_incomplete_type` or whichever).

File: toolchain/check/import_ref.cpp
Comment (chandlerc) [inline toolchain/check/import_ref.cpp]: And here, since we're also manipulating the member of `Work`, I wonder if there is a different name for the function variable to indicate the separation?

`already_resolved_incomplete_type` is awfully long.

Alternatively, could consider a different naming structure: `pre_incomplete_type` to indicate the first phase?

Or to keep `false`-by-default, maybe `type_second_pass`? And then these comments can remind the reader that the second pass doesn't need to create the incomplete type but does need to load. WDYT? Happy to chat about naming options if none of these work.

**Review by chandlerc — APPROVED** [review]
The multi-phase processing I think was what was tripping me up. Thanks for all the comments and improvements here, I think this makes sense to me now.

The change LG structurally. Some trivial stuff inline and a test fix, LGTM with those in place.

I've also left some comments discussing some different options for naming the bool used to track the two passes here. Happy for you to merge with whatever names seem best to you (even the current ones) and continue that discussion in Discord or with more folks to see if there is a better name. That's a somewhat minor tweak to maybe make things read a bit better in the future, but I don't want to block on it so approving. =]

File: toolchain/check/import_ref.cpp
Comment (jonmeow) [reply toolchain/check/import_ref.cpp:44]: Restructuring with "For most cases:" to make the intended overlap clear.

File: toolchain/check/import_ref.cpp
Comment (jonmeow) [reply toolchain/check/import_ref.cpp]: Rephrased as discussed.

File: toolchain/check/import_ref.cpp
Comment (jonmeow) [reply toolchain/check/import_ref.cpp]: Renamed to `made_incomplete_type`, hoping it helps (`resolved` just implies to me a more complete state)

File: toolchain/check/import_ref.cpp
Comment (jonmeow) [reply toolchain/check/import_ref.cpp]: Done

File: toolchain/check/import_ref.cpp
Comment (jonmeow) [reply toolchain/check/import_ref.cpp]: As above, `made_incomplete_type`

`type_second_pass` isn't quite the semantic I was aiming for because we still might need to check the constant to see if the instruction was previously filled in for some other edge cases.

File: toolchain/check/import_ref.cpp
Comment (jonmeow) [reply toolchain/check/import_ref.cpp]: Done

**Review by jonmeow — COMMENTED** [review]
Per discussion, also added a struct cycle test.

File: toolchain/check/testdata/class/import_member_cycle.carbon
Comment (jonmeow) [reply toolchain/check/testdata/class/import_member_cycle.carbon]: Done

File: toolchain/check/import_ref.cpp
Comment (chandlerc) [inline toolchain/check/import_ref.cpp]: WDYT:

```suggestion
//        iii. Otherwise, continue resolving but set made_incomplete_type.
```

**Review by chandlerc — APPROVED** [review]
Niiiiice!

**Pull request merged**
